File: app/src/main/java/life/plank/juna/zone/view/adapter/SearchViewAdapter.java
package life.plank.juna.zone.view.adapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.User;
import life.plank.juna.zone.interfaces.OnItemClickListener;

public class SearchViewAdapter extends RecyclerView.Adapter<SearchViewAdapter.SearchViewHolder> {

    private List<User> userList;
    private Context context;
    private Picasso picasso;
    private OnItemClickListener onItemClickListener;

    public SearchViewAdapter(List<User> userList, Context context, OnItemClickListener onItemClickListener, Picasso picasso) {
        this.context = context;
        this.userList = userList;
        this.onItemClickListener = onItemClickListener;
        this.picasso = picasso;
    }

    @Override
    public SearchViewAdapter.SearchViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new SearchViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_image_and_title, parent, false));
    }

    @Override
    public void onBindViewHolder(SearchViewAdapter.SearchViewHolder holder, int position) {
        holder.usernameTextView.setText(userList.get(position).getDisplayName());
        picasso.load(userList.get(position).getProfilePictureUrl())
                .placeholder(R.drawable.ic_default_profile)
                .error(R.drawable.ic_default_profile)
                .into(holder.profileImageView);
        holder.profileImageView.setOnClickListener(view -> {
            if (holder.followTick.getVisibility() == View.VISIBLE) {
                holder.followTick.setVisibility(View.INVISIBLE);
                holder.profileImageView.setAlpha(255);
                holder.usernameTextView.setVisibility(View.VISIBLE);
                holder.usernameTextView.setTextColor(context.getResources().getColor(R.color.grey));
                holder.profileImageView.clearColorFilter();
                onItemClickListener.onItemClicked(userList.get(position).getObjectId(), false);
            } else {
                holder.followTick.setVisibility(View.VISIBLE);
                holder.profileImageView.setAlpha(160);
                holder.usernameTextView.setTextColor(Color.BLACK);
                holder.profileImageView.setColorFilter(context.getResources().getColor(R.color.red_pink), PorterDuff.Mode.LIGHTEN);
                onItemClickListener.onItemClicked(userList.get(position).getObjectId(), true);
            }
        });
    }

    @Override
    public int getItemCount() {
        return userList.size();
    }

    public void update(List<User> users) {
        userList.clear();
        this.userList.addAll(users);
        notifyDataSetChanged();
    }

    public class SearchViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.image)
        ImageView profileImageView;
        @BindView(R.id.title)
        TextView usernameTextView;
        @BindView(R.id.follow_image_view)
        ImageView followTick;

        SearchViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/view/adapter/common/SearchViewAdapter.kt
package life.plank.juna.zone.view.adapter.common

import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.item_image_and_title.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.data.model.User
import life.plank.juna.zone.util.facilis.onDebouncingClick
import life.plank.juna.zone.util.view.UIDisplayUtil.findColor

class SearchViewAdapter(private val userList: MutableList<User>, private val glide: RequestManager) : RecyclerView.Adapter<SearchViewAdapter.SearchViewHolder>() {

    val selectedUsers: MutableSet<User> = HashSet()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchViewHolder =
            SearchViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_image_and_title, parent, false))

    override fun onBindViewHolder(holder: SearchViewHolder, position: Int) {
        holder.itemView.run {
            val user = userList[position]
            title.text = user.displayName
            glide.load(user.profilePictureUrl)
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_default_profile)
                            .error(R.drawable.ic_default_profile))
                    .into(image)

            setItemSelection(user, holder, false)

            image.onDebouncingClick { setItemSelection(user, holder, true) }
        }
    }

    private fun setItemSelection(user: User, holder: SearchViewHolder, isActionDone: Boolean) {
        holder.itemView.run {
            if (isActionDone) {
                if (selectedUsers.contains(user)) {
                    selectedUsers.remove(user)
                } else {
                    selectedUsers.add(user)
                }
            }
            val isItemSelected = selectedUsers.contains(user)
            follow_image_view.visibility = if (isItemSelected) View.VISIBLE else View.INVISIBLE
            image.imageAlpha = if (isItemSelected) 160 else 255
            title.visibility = if (isItemSelected) View.INVISIBLE else View.VISIBLE
            title.setTextColor(if (isItemSelected) Color.BLACK else findColor(R.color.grey))
            image.colorFilter = (if (isItemSelected) PorterDuffColorFilter(findColor(R.color.red_pink), PorterDuff.Mode.LIGHTEN) else null)
        }
    }

    override fun getItemCount(): Int = userList.size

    fun update(users: List<User>) {
        userList.clear()
        this.userList.addAll(users)
        notifyDataSetChanged()
    }

    class SearchViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}
